File: 2020/day17/sol.cpp
#include <array>
#include <fstream>
#include <iostream>
#include <tuple>

using v4 = std::tuple<int, int, int, int>;
using Grid4d = std::array<std::array<std::array<std::array<char, 20>, 20>, 13>, 13>;

int get_nearby_cubes(const Grid4d& grid, v4 coords, bool four_dim = false) {
    auto [x, y, z, w] = coords;

    int from = four_dim ? w - 1 : w;
    int to = four_dim ? w + 1 : w;

    int count = 0;
    for (int nw = from; nw <= to; ++nw) {
        for (int nz = z - 1; nz <= z + 1; ++nz) {
            for (int ny = y - 1; ny <= y + 1; ++ny) {
                for (int nx = x - 1; nx <= x + 1; ++nx) {
                    if (nz == z && ny == y && nx == x && nw == w) {
                        continue;
                    }
                    if (nw < 0 || nz < 0 || ny < 0 || nx < 0) {
                        continue;
                    }
                    if (   nw >= std::ssize(grid)
                        || nz >= std::ssize(grid[0])
                        || ny >= std::ssize(grid[0][0])
                        || nx >= std::ssize(grid[0][0][0])) {
                        continue;
                    }
                    if (grid[nw][nz][ny][nx] == '#') {
                        count++;
                    }
                }
            }
        }
    }

    return count;
}

Grid4d cycle(const Grid4d& current_state, bool four_dim = false) {
    Grid4d next_state = {};

    int from = four_dim ? 0 : 6;
    int to = four_dim ? std::ssize(current_state) : 7;

    for (int w = from; w < to; ++w) {
        for (int z = 0; z < std::ssize(current_state[0]); ++z) {
            for (int y = 0; y < std::ssize(current_state[0][0]); ++y) {
                for (int x = 0; x < std::ssize(current_state[0][0][0]); ++x) {
                    int nearby = get_nearby_cubes(current_state, { x, y, z, w }, four_dim);
                    auto&& next_cube = next_state[w][z][y][x];
                    if (current_state[w][z][y][x] == '#') {
                        if (2 <= nearby && nearby <= 3) {
                            next_cube = '#';
                        } else {
                            next_cube = '.';
                        }
                    } else if (nearby == 3) {
                        next_cube = '#';
                    } else {
                        next_cube = '.';
                    }
                }
            }
        }
    }

    return next_state;
}

int count_active(const Grid4d& grid, bool four_dim = false) {
    int count = 0;

    int from = four_dim ? 0 : 6;
    int to = four_dim ? std::ssize(grid) : 7;

    for (int w = from; w < to; ++w) {
        for (int z = 0; z < std::ssize(grid[0]); ++z) {
            for (int y = 0; y < std::ssize(grid[0][0]); ++y) {
                for (int x = 0; x < std::ssize(grid[0][0][0]); ++x) {
                    if (grid[w][z][y][x] == '#') {
                        count++;
                    }
                }
            }
        }
    }
    return count;
}

int simulate(Grid4d grid, bool four_dim = false) {
    for (int i = 0; i < 6; ++i) {
        grid = cycle(grid, four_dim);
    }

    return count_active(grid, four_dim);
}

int main() {
    std::ifstream input("input");
    if (!input.is_open()) {
        std::cerr << "Can't read puzzle input file.\n";
        return -1;
    }

    Grid4d grid = {};
    {
        std::string line;
        for (int y = 0; std::getline(input, line); ++y) {
            for (size_t x = 0; x < line.size(); ++x) {
                grid[6][6][y + 6][x + 6] = line[x];
            }
        }
    }
    input.close();

    std::cout << "part 1: " << simulate(grid) << '\n';
    std::cout << "part 2: " << simulate(grid, true) << '\n';

    return 0;
}

File: 2020/day17/sol.rs
use std::fs;

type V4 = (i32, i32, i32, i32);
type Grid4d = [[[[u8; 20]; 20]; 13]; 13];

fn get_nearby_cubes(grid: &Grid4d, (x, y, z, w): V4, four_dim: bool) -> usize {
    let mut count = 0;

    let (from, to) = if four_dim { (-1, 1) } else { (0, 0) };
    for dw in from..=to {
        for dz in -1..=1 {
            for dy in -1..=1 {
                for dx in -1..=1 {
                    if (dw, dz, dy, dx) == (0, 0, 0, 0) {
                        continue;
                    }

                    let (nw, nz, ny, nx) = (w + dw, z + dz, y + dy, x + dx);
                    if nw < 0 || nz < 0 || ny < 0 || nx < 0 {
                        continue;
                    }

                    let (nw, nz, ny, nx) = (nw as usize, nz as usize, ny as usize, nx as usize);
                    if nw >= grid.len()
                        || nz >= grid[0].len()
                        || ny >= grid[0][0].len()
                        || nx >= grid[0][0][0].len()
                    {
                        continue;
                    }

                    if grid[nw][nz][ny][nx] == b'#' {
                        count += 1;
                    }
                }
            }
        }
    }

    count
}

fn cycle(grid: &Grid4d, four_dim: bool) -> Grid4d {
    let mut next_state: Grid4d = [[[[0; 20]; 20]; 13]; 13];

    let (from, to) = if four_dim { (0, grid.len()) } else { (6, 7) };
    for w in from..to {
        for z in 0..grid[0].len() {
            for y in 0..grid[0][0].len() {
                for x in 0..grid[0][0][0].len() {
                    let nearby =
                        get_nearby_cubes(grid, (x as i32, y as i32, z as i32, w as i32), four_dim);

                    let current_cube = grid[w][z][y][x];
                    let next_cube = &mut next_state[w][z][y][x];

                    if current_cube == b'#' {
                        if (2..=3).contains(&nearby) {
                            *next_cube = b'#';
                        } else {
                            *next_cube = b'.';
                        }
                    } else if nearby == 3 {
                        *next_cube = b'#';
                    } else {
                        *next_cube = b'.';
                    }
                }
            }
        }
    }

    next_state
}

fn count_active(grid: &Grid4d, four_dim: bool) -> usize {
    // if we count only in 3 dimensions, we can skip other 3d cubes
    let (from, to) = if four_dim { (0, grid.len()) } else { (6, 7) };

    grid.iter()
        .skip(from)
        .take(to - from)
        .flatten()
        .flatten()
        .flatten()
        .filter(|&c| c == &b'#')
        .count()
}

fn simulate(mut grid: Grid4d, four_dim: bool) -> usize {
    for _ in 0..6 {
        grid = cycle(&grid, four_dim);
    }

    count_active(&grid, four_dim)
}

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let mut grid: Grid4d = [[[[0; 20]; 20]; 13]; 13];
    input.lines().enumerate().for_each(|(y, line)| {
        line.chars().enumerate().for_each(|(x, c)| {
            grid[6][6][y + 6][x + 6] = c as u8;
        });
    });

    println!("part 1: {}", simulate(grid, false));
    println!("part 2: {}", simulate(grid, true));
}
